subsys/bluetooth/mesh/vnd/dm_srv.c

```diff
@@ -0,0 +1,474 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <bluetooth/mesh/vnd/dm_srv.h>
+#include "../model_utils.h"
+#include <dm.h>
+#include "mesh/net.h"
+#include "mesh/access.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_MODEL)
+#define LOG_MODULE_NAME bt_mesh_dm_srv
+#include "common/log.h"
+
+struct bt_mesh_dm_srv *dm_srv;
+
+struct settings_data {
+	struct bt_mesh_dm_cfg cfg;
+} __packed;
+
+#if CONFIG_BT_SETTINGS
+static void store_timeout(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_mesh_dm_srv *srv = CONTAINER_OF(
+		dwork, struct bt_mesh_dm_srv, store_timer);
+
+	struct settings_data data = {
+		.cfg = srv->cfg,
+	};
+
+	(void)bt_mesh_model_data_store(srv->model, true, NULL, &data,
+				       sizeof(data));
+}
+#endif
+
+static void store(struct bt_mesh_dm_srv *srv)
+{
+#if CONFIG_BT_SETTINGS
+	k_work_schedule(
+		&srv->store_timer,
+		K_SECONDS(CONFIG_BT_MESH_MODEL_SRV_STORE_TIMEOUT));
+#endif
+}
+
+static uint16_t result_convert_cm(float val)
+{
+	int32_t temp = val * 100;
+
+	if (temp < 0) {
+		return 0;
+	}
+
+	return temp > UINT16_MAX ? UINT16_MAX : (uint16_t)temp;
+}
+
+static void new_entry_store(struct dm_result *result)
+{
+	dm_srv->results.last_entry_idx =
+		(dm_srv->results.last_entry_idx + 1) % dm_srv->results.entry_cnt;
+
+	struct bt_mesh_dm_res_entry *entry = &dm_srv->results.res[dm_srv->results.last_entry_idx];
+
+	dm_srv->results.available_entries =
+		(dm_srv->results.available_entries + 1) >= dm_srv->results.entry_cnt ?
+			dm_srv->results.entry_cnt :
+			(dm_srv->results.available_entries + 1);
+
+	memset(entry, 0, sizeof(struct bt_mesh_dm_res_entry));
+
+	entry->method = result->ranging_mode;
+	entry->addr = dm_srv->target_addr;
+	entry->quality = result->quality;
+	entry->err_occurred = !result->status;
+
+	if (entry->err_occurred) {
+		return;
+	}
+
+	if (entry->method == DM_RANGING_MODE_RTT) {
+		entry->res.rtt = result_convert_cm(result->dist_estimates.rtt.rtt);
+	} else {
+		entry->res.mcpd.best = result_convert_cm(result->dist_estimates.mcpd.best);
+		entry->res.mcpd.ifft = result_convert_cm(result->dist_estimates.mcpd.ifft);
+		entry->res.mcpd.phase_slope =
+			result_convert_cm(result->dist_estimates.mcpd.phase_slope);
+		entry->res.mcpd.rssi_openspace =
+			result_convert_cm(result->dist_estimates.mcpd.rssi_openspace);
+	}
+}
+
+static void cfg_status_send(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *rx_ctx,
+			    uint8_t status)
+{
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	BT_MESH_MODEL_BUF_DEFINE(msg, BT_MESH_DM_CLI_OP,
+				 BT_MESH_DM_CONFIG_STATUS_MSG_LEN);
+	bt_mesh_model_msg_init(&msg, BT_MESH_DM_CLI_OP);
+	net_buf_simple_add_u8(&msg, BT_MESH_DM_CONFIG_STATUS_OP);
+	net_buf_simple_add_u8(&msg, status);
+	net_buf_simple_add_u8(&msg, (srv->is_busy << 7) |
+					    (srv->results.available_entries & 0x7F));
```

Comment:
So apparently the max usable array size is 127, this should a) be mentioned somewhere b) perhaps be enforced somehow by the API.

Also: should we change the spec here to store `available entries - 1` in the state? So we can use 128 entries, having 0 does not make much sense anyways so we don't need to support it.

Reply:
On second thought: This is also limited by the amount of bytes you can fit into a results message. The range of sensible sizes (>= 1, <= max entries that can fit in msg) should definitely be mentioned somewhere, and ideally be enforced in code.

Reply:
If we are going to do this I would prefer to do it along with the other documentation improvments to prevent stalling this PR

Reply:
Backlog item.